### PR TITLE
RTL Conversion - NxTag - RSC-1352

### PR DESCRIPTION
**Pull request by avnvbhatta**
JIRA: https://issues.sonatype.org/browse/RSC-1352

### CODE REVIEW

File: lib/src/components/NxTag/__tests__/NxTag.test.tsx [hunk 1]
```diff
@@ -4,124 +4,125 @@
...
-  const getShallowComponent = enzymeUtils.getShallowComponent(NxTag, { children: 'basic tag' });
+  type PropsWithRef = PublicProps & RefAttributes<HTMLLabelElement>;
+  const renderEl = rtlRenderElement<PropsWithRef>(NxTag, { children: 'basic tag '});
 
   it('renders NxTag with the `nx-tag` class', function() {
-    expect(getShallowComponent().children().find('.nx-tag')).toExist();
+    const el = renderEl()!;
+    expect(el.tagName).toBe('LABEL');
+    expect(el).toHaveClass('nx-tag');
   });
```
Comment (rpokorny) [inline lib/src/components/NxTag/__tests__/NxTag.test.tsx:21]: IMO both the classname and the tagName are impl details

File: lib/src/components/NxTag/__tests__/NxTag.test.tsx [hunk 2]
```diff
@@ -4,124 +4,125 @@
...
 
   it('correctly assigns supplied id', function() {
-    const idComponent = getShallowComponent({ id: 'test-id' });
-    expect(idComponent.find('.nx-tag')).toHaveProp('id', 'test-id');
+    const el = renderEl({ id: 'test-id' });
+    expect(el).toHaveAttribute('id', 'test-id');
   });
 
   it('renders the supplied text inside .nx-tag__text', function() {
```
Comment (rpokorny) [inline lib/src/components/NxTag/__tests__/NxTag.test.tsx:46]:
```suggestion
  it('renders the supplied text', function() {
```

File: lib/src/components/NxTag/__tests__/NxTag.test.tsx [hunk 3]
```diff
@@ -4,124 +4,125 @@
...
+    expect(el).toHaveTextContent('tag text');
   });
 
   it('sets the nx-selectable-color--indigo class if no color prop is passed', function() {
-    expect(getShallowComponent().find('.nx-selectable-color--indigo')).toExist();
+    const el = renderEl();
+    expect(el).toHaveClass('nx-selectable-color--indigo');
   });
 
   it('sets the color class using the color if it is provided', function() {
```
Comment (rpokorny) [inline lib/src/components/NxTag/__tests__/NxTag.test.tsx:56]: Should be tested via visual tests; the classnames are impl details

File: lib/src/components/NxTag/__tests__/NxTag.test.tsx [hunk 4]
```diff
@@ -4,124 +4,125 @@
...
+    expect(el).toHaveClass('nx-selectable-color--orange');
   });
 });
 
 describe('NxSelectableTag', function() {
-  const minimumProps = { children: 'selectable tag', selected: false, onSelect: jest.fn() };
-  const getShallowComponent = enzymeUtils.getShallowComponent(NxSelectableTag, minimumProps);
+  type PropsWithRef = SelectableProps & RefAttributes<HTMLLabelElement>;
+
+  const minimalProps = { children: 'selectable tag', selected: false, onSelect: jest.fn() },
```
Comment (rpokorny) [inline lib/src/components/NxTag/__tests__/NxTag.test.tsx:65]: Spies should only be created per-test (either in `it` or `beforeEach` blocks).  Our typical pattern for `minimalProps` is to fill in required functions with a simple noop, and then to replace it with a spy in the tests that actually need it
```suggestion
  const minimalProps = { children: 'selectable tag', selected: false, onSelect: () => {} },
```

File: lib/src/components/NxTag/__tests__/NxTag.test.tsx [hunk 5]
```diff
@@ -4,124 +4,125 @@
...
+      quickRender = rtlRender<PropsWithRef>(NxSelectableTag, minimalProps),
+      renderEl = rtlRenderElement<PropsWithRef>(NxSelectableTag, minimalProps);
 
   it('renders an NxTag', function() {
-    expect(getShallowComponent()).toMatchSelector('ForwardRef(NxTag)');
+    const el = renderEl()!;
+    expect(el.tagName).toBe('LABEL');
+    expect(el).toHaveClass('nx-tag');
   });
```
Comment (rpokorny) [inline lib/src/components/NxTag/__tests__/NxTag.test.tsx:73]: Impl details

File: lib/src/components/NxTag/__tests__/NxTag.test.tsx [hunk 6]
```diff
@@ -4,124 +4,125 @@
...
+    expect(icon).toHaveClass('fa-plus-circle nx-tag__action');
   });
 
   it('renders the times icon when selected', function() {
-    const getSelectedComponent = enzymeUtils.getShallowComponent(NxSelectableTag,
-        { children: 'selectable tag', selected: true, onSelect: jest.fn() });
+    const el = quickRender({ selected: true }),
+        icon = el.getByRole('img', { hidden: true });
+    expect(icon).toHaveClass('fa-times-circle nx-tag__action');
+  });
```
Comment (rpokorny) [inline lib/src/components/NxTag/__tests__/NxTag.test.tsx:106]: This all seems like impl details that should be covered in visual tests

File: lib/src/components/NxTag/__tests__/NxTag.test.tsx [hunk 7]
```diff
@@ -4,124 +4,125 @@
...
+        tagEl = el.container.querySelector('.nx-tag'),
+        textEl = el.container.querySelector('span'),
+        iconEl = el.getByRole('img', { hidden: true });
 
-    expect(icon).toMatchSelector(NxFontAwesomeIcon);
-    expect(icon).toHaveProp(icon, faTimesCircle);
-    expect(icon).toHaveClassName('nx-tag__action');
+    expect(tagEl?.firstChild).toContainElement(textEl);
+    expect(tagEl?.lastChild).toContainElement(iconEl);
   });
```
Comment (rpokorny) [inline lib/src/components/NxTag/__tests__/NxTag.test.tsx:116]: Also impl/style details

File: lib/src/components/NxTag/__tests__/NxTag.test.tsx [hunk 8]
```diff
@@ -4,124 +4,125 @@
...
 
-  it('fires the components onSelect when clicked', function() {
-    const onSelect = jest.fn(),
-        trigger = getShallowComponent({ onSelect });
+  it('fires the components onSelect when clicked', async function() {
+    const user = userEvent.setup(),
+        onSelect = jest.fn(),
+        { container } = quickRender({ onSelect });
 
+    const tag = container.querySelector<HTMLElement>('label')!;
```
Comment (rpokorny) [inline lib/src/components/NxTag/__tests__/NxTag.test.tsx:123]:
```suggestion
        tag = renderEl({ onSelect });

```

File: lib/src/components/NxTag/__tests__/NxTag.test.tsx (hunk 1 above)
Comment (avnvbhatta) [reply lib/src/components/NxTag/__tests__/NxTag.test.tsx]: > IMO both the classname and the tagName are impl details

Removed here: 607f83a034413c21cc13758c8e42ec8a7c0ba678

File: lib/src/components/NxTag/__tests__/NxTag.test.tsx (hunk 3 above)
Comment (avnvbhatta) [reply lib/src/components/NxTag/__tests__/NxTag.test.tsx]: > Should be tested via visual tests; the classnames are impl details

Changed here: 4e903318fd82ddafd673824edd89ea5bee4d6d89

File: lib/src/components/NxTag/__tests__/NxTag.test.tsx (hunk 5 above)
Comment (avnvbhatta) [reply lib/src/components/NxTag/__tests__/NxTag.test.tsx]: > Impl details

Changed here: 00e685e8af9a11fce666fe07fd7d52e8aee8fb3a

File: lib/src/components/NxTag/__tests__/NxTag.test.tsx (hunk 6 above)
Comment (avnvbhatta) [reply lib/src/components/NxTag/__tests__/NxTag.test.tsx]: > This all seems like impl details that should be covered in visual tests

I removed all of the tests that checked for classes. a51ebe2fee38cc33d49fa21c337977c5e47bee55

File: lib/src/components/NxTag/__tests__/NxTag.test.tsx (hunk 7 above)
Comment (avnvbhatta) [reply lib/src/components/NxTag/__tests__/NxTag.test.tsx]: > Also impl/style details

Removed here: a51ebe2fee38cc33d49fa21c337977c5e47bee55

File: gallery/visualtests/nxTag.js [hunk 9]
```diff
@@ -57,6 +60,21 @@ describe('NxTag', function() {
...
       await checkScreenshot(targetElement);
     });
+
+    it('has a plus icon when not selected', async function() {
+      const [targetElement] = await waitAndGetElements(selector);
+
+      await checkScreenshot(targetElement);
+    });
+
+    it('has a times icon when selected', async function() {
```
Comment (rpokorny) [inline gallery/visualtests/nxTag.js:70]: Aren't these icons covered in the screenshots of the "by default" and "when clicked" tests above?

**Review by rpokorny — APPROVED** [review]

File: gallery/visualtests/nxTag.js (hunk 9 above)
Comment (avnvbhatta) [reply gallery/visualtests/nxTag.js]: > Aren't these icons covered in the screenshots of the "by default" and "when clicked" tests above?

Oh yes, removed the redundant test here: c153bbd5e4583c645d8ad17368aa71494ab126b4

File: lib/src/components/NxTag/__tests__/NxTag.test.tsx [hunk 10]
```diff
@@ -4,124 +4,69 @@
...
-
-    expect(getSelectedComponent.find('.nx-tag').children().first()).toMatchSelector('span');
-    expect(getSelectedComponent.find('.nx-tag').children().last()).toHaveClassName('nx-tag__action');
+  it('renders the supplied text', function() {
+    const el = renderEl({ children: 'tag text' });
+    expect(el).toHaveTextContent('tag text');
   });
 });
 
 describe('NxSelectableTag', function() {
```
Comment (rpokorny) [inline lib/src/components/NxTag/__tests__/NxTag.test.tsx:46]: `NxSelectableTag`s include a checkbox with the `switch` role and who's `checked` attr reflects the `selected` prop.  That should all be tested

Comment (avnvbhatta) [reply lib/src/components/NxTag/__tests__/NxTag.test.tsx:46]: > `NxSelectableTag`s include a checkbox with the `switch` role and who's `checked` attr reflects the `selected` prop. The checkbox's a11y name is equal to the tag text. That should all be tested

Added here: ad6e6ee522d01288b3b2dbc9154bb640228bb787

**Pull request merged**